### PR TITLE
[WIP] Refactor to Introduce Backend Abstraction

### PR DESCRIPTION
**Pull request by zhenglongjiepheonix**
# What does this PR do?

- [x] add backend abstraction
- [x] refactor the original pipeline flow to accommodate potential needs of different backend
- [x] modify API so that more parameter passing format will be supported

The `NanotronBackend` is still WIP and untested, but it would be nice to get some feedbacks first.

### CODE REVIEW

File: optimum/fx/parallelization/api.py [hunk 1]
```diff
@@ -34,32 +34,40 @@
...
 
     Args:
-        model (`str`):
-            Model to parallelize, a model id on the Huggingface Hub or path to a local directory containing config and weights
-            of the model.
         parallel_ctx (`ParallelExecutionCtx`):
             Parallel execution context containing process groups the current process belongs to.
         *model_args (`Any`):
             Additional postional arguments for intializing the model if a model id is passed.
+        model_id_or_path (`str`):
```
Comment (michaelbenayoun) [inline optimum/fx/parallelization/api.py:64]:
```suggestion
        model_id_or_path (`Optional[str]`, defaults to `None`):
```

**Review by michaelbenayoun — COMMENTED** [review]
Very nice first draft.

I was wondering: do you think we should keep the 2 methods for row and column separated? Just opening the question, I do not have a strong opinion on that.

File: optimum/fx/parallelization/backend/base.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+    def create_parallel_embedding(
+        self,
+        mod: nn.Embedding,
+        parallel_ctx: "ParallelExecutionCtx",
+        sequence_parallel: bool,
+        contiguous_chunks: Optional[Tuple[int]] = None,
+    ) -> nn.Module:
+        raise NotImplementedError
+
+    def pre_process(self, graph_module: GraphModule, ctx: "ParallelExecutionCtx", config: "Config") -> GraphModule:
```
[the remark this answers is not in the document]
Comment (zhenglongjiepheonix) [reply optimum/fx/parallelization/backend/base.py:102]: a `Config` is a data class which records static configurations during the whole process

File: optimum/fx/parallelization/backend/base.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        """
+        Mark tie information right before we run passes because dynamo tracing will alter the parameter name while our
+        passes don't.
```
[the remark this answers is not in the document]
Comment (zhenglongjiepheonix) [reply optimum/fx/parallelization/backend/base.py]: I think generally it means something that needs done before we run passes, and weights tying happens to be one of them

**Comment by michaelbenayoun** [comment]
@zhenglongjiepheonix what's the status?

**Comment by zhenglongjiepheonix** [comment]
the nanotron backend is not tested, for the default backend, everything works fine, it contains the newest update and has addressed comments

**Comment by michaelbenayoun** [comment]
Ok so ready for final review?

**Comment by zhenglongjiepheonix** [comment]
> Ok so ready for final review?

Basically It's for reference, if someone is working on the support for nanotron indeed, then the correctness of nanotronbackend needs to be verified and additional tests are needed, but in my opinion this PR marks the boundary of optimum and nanotron, the rest of work should be implemented inside nanotron and optimum just expose `parallelize_model` api
